### PR TITLE
Improve a DX issue with IProcessedStyleSet typings. (issue #6124)

### LINKED ISSUE #6124: Can't seem to style Button through 'subComponentStyles' in IProcessedStyleSet<>
Now that #6061 landed I can no longer apply styling to Buttons through the `subComponentStyles` property on my `IProcessedStyleSet<>` without jumping through extra hoops:

If I do this:
`styles={this._classNames.subComponentStyles.button}` 
I get an error:
```
Type 'IStyleFunction<{}, IButtonStyles> | (() => Partial<IButtonStyles>)' is not assignable to type 'IButtonStyles | undefined'.
  Value of type 'IStyleFunction<{}, IButtonStyles>' has no properties in common with type 'IButtonStyles'. Did you mean to call it?
```
So I try to call it:
`styles={this._classNames.subComponentStyles.button({} /* empty as there is no IButtonStyleProps */)}`
or like this:
`styles={this._classNames.subComponentStyles.button()}`

and I get a different error:
```
[ts] Cannot invoke an expression whose type lacks a call signature. Type 'IStyleFunction<{}, IButtonStyles> | (() => Partial<IButtonStyles>)' has no compatible call signatures.
```
So it seems I need some kind of construct like this now:
`styles={(this._classNames.subComponentStyles.button as (IStyleFunction<{}, IButtonStyles>))({})}`

But if I read the types correctly couldn't `IStyleFunction<IStyleProps, IStyles> | (() => Partial<IStyles>)` be made more succinct if `IStyleFunction` was defined like this (with "props" being optional):
```
export declare type IStyleFunction<TStylesProps, TStyleSet extends IStyleSet<TStyleSet>> = (props?: TStylesProps) => Partial<TStyleSet>;
```
But this might have consequences elsewhere, and I haven't yet tried to apply the change to see what those might be so it might not be the way to go. Though, if the `props` argument really is optional perhaps it is a suitable fix, and if it isn't, could we remove `(() => Partial<IStyles>)` from the type definition? I'm not too familiar with the internals here so not sure why this definition looks the way it does in the first place, but maybe there is a good reason I'm not seeing?

### PR DESCRIPTION
**Pull request by cliffkoh**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #6124
- [x] Include a change request file using `$ npm run change`

#### Description of changes

Addresses the main problem in #6124. This is not a perfect fix - the perfect fix would require consumers be on Typescript 3. For that matter, it requires OUFR itself to be on Typescript 3 so that our test cases pass.

This however will work for what is likely to be the far more common pattern (the first two test cases). For the edge cases (the last 2 cases), it should be fortunately relatively easy to workaround the problem.

#### Focus areas to test

(optional)

### CODE REVIEW

**Comment by cliffkoh** [comment]
@kenotron Issue #6147 also tracks a follow-up fix for this in the future.

File: packages/merge-styles/src/IStyleSet.ts [hunk 1]
```diff
@@ -6,7 +6,18 @@ export type Diff<T extends keyof any, U extends keyof any> = ({ [P in T]: P } &
...
 
-export type TrimToFunction<T> = T extends (...args: any[]) => any ? T : () => T;
+export type __ReduceToFunction<T> = T extends (...args: any[]) => any ? T : never;
+
+/**
+ * Helper function whose role is supposed to express that regardless if T is a style object or style function,
+ * it will always map to a style function.
+ *
+ * Note: the commmented segment is the right expression but it is buggy in 2.8.2. Once Fabric upgrades to Typescript 3,
+ *  we should uncomment the commented segmment below as well as uncomment the test in mergeStyleSets.test.ts.
```
Comment (JasonGore) [inline packages/merge-styles/src/IStyleSet.ts]: I wonder if we should flag these with something we can search for later, like `__TS_2_8_WORKAROUND__` (which is inconsistently named already.)

**Comment by cliffkoh** [comment]
It is not just a comment :D

File: packages/merge-styles/src/mergeStyleSets.test.ts [hunk 2]
```diff
@@ -196,14 +195,118 @@ describe('mergeStyleSets', () => {
...
-    const styleSet2: ITestClasses = mergeStyleSets(styleSet1, {
+    const styleSet2 = mergeStyleSets(styleSet1, {
       root: ['ms-Bar', { background: 'green' }]
     });
 
     expect(styleSet2).toEqual({ root: 'ms-Foo ms-Bar root-1', subComponentStyles: {} });
     expect(_stylesheet.getRules()).toEqual('.root-0{background:red;}' + '.root-1{background:green;}');
   });
+
+  describe('typings tests', () => {
```
Comment (JasonGore) [inline packages/merge-styles/src/mergeStyleSets.test.ts]: 👍

File: packages/merge-styles/src/IStyleSet.ts (hunk 1 above)
Comment (cliffkoh) [reply packages/merge-styles/src/IStyleSet.ts]: There is issue #6147 which explicitly tracks this todo, assigned to me. I think that is enough since I scrub my assigned items regularly but let me know if you strongly feel I should add the "search term" in the comment.

File: packages/merge-styles/src/mergeStyleSets.test.ts [hunk 3]
```diff
@@ -196,14 +195,118 @@ describe('mergeStyleSets', () => {
...
+      }
+    });
+
+    it('IStyleSet/IProcessedStyleSet should work with standard sub components', () => {
+      const classNames = mergeStyleSets<IStyles>(getStyles());
+
+      SubComponent({ styles: classNames.subComponentStyles.button });
+
+      // this test primarily tests that the lines above do not result in a Typescript error.
+      expect.assertions(0);
```
Comment (JasonGore) [inline packages/merge-styles/src/mergeStyleSets.test.ts]: Could `it` just be renamed so it's not an executable test case while still having typings generate an error?

Comment (kenotron) [reply packages/merge-styles/src/mergeStyleSets.test.ts]: This is to exercise compiler though...

Comment (cliffkoh) [reply packages/merge-styles/src/mergeStyleSets.test.ts]: Renamed how? I actually would like it to show up as an executable test actually that reflects in the jest log, and also unit test count, since these are actually testing things :)

File: packages/merge-styles/src/IStyleSet.ts (hunk 1 above)
Comment (JasonGore) [reply packages/merge-styles/src/IStyleSet.ts]: Oops, I didn't see that comment earlier.

File: packages/merge-styles/src/mergeStyleSets.test.ts (hunk 3 above)
Comment (JasonGore) [reply packages/merge-styles/src/mergeStyleSets.test.ts]: @kenotron These tests would fail the build without Jest actually running them, but I see Cliff's point about stat tracking
